### DISCUSSION TITLE
Building the scylladb-php-driver on an Apple M3 Pro (ARM64 Architecture)

### DISCUSSION THREAD

**Discussion by latublu** [nano-interactive/cassandra-php-driver#92] [Q&A]
I'm attempting to build the driver on an Apple M3 Pro, and I am in need of advice.

# TL;DR
> My ultimate goal is to setup a dev environment for a PHP project that needs to utilize a ScyllaDB server as a data source. My preferred work station is a Mac. 
> Is it possible to build the driver on ARM64 Architecture?
> Is anyone using the driver on Apple M3 Pro?
> What are your build or installation steps for your tech stack?
> Is there a better approach for a PHP project that needs to utilize a ScyllaDB server as a data source?

# My Build Attempts

By initial efforts in building it directly on an Apple M3 Pro were not successful.

(I was successful in building the driver in Ubuntu 22.04 LTS that is running on x86-64 Architecture.)

My current attempt is to build the driver on  aUbuntu 22.04 LTS virtual machine, running on a Apple M3 Pro. I'm running the VM using UTM.

I followed the procedures in the repo README file, and I have documented my exact procedure.

https://latu.notion.site/Build-Procedure-for-ScyllaDB-PHP-Driver-e9c9b88b906040f789d82b47e6c837f3

The build process did appear to finish, and I completed the installation steps.

But I have a block. The following startup error is being thrown by PHP.

```
# php -m
PHP Warning:  PHP Startup: Unable to load dynamic library 'cassandra.so' (tried: /usr/lib/php/20210902/cassandra.so (libscylla-cpp-driver.so.2: cannot open shared object file: No such file or directory), /usr/lib/php/20210902/cassandra.so.so (/usr/lib/php/20210902/cassandra.so.so: cannot open shared object file: No such file or directory)) in Unknown on line 0
```

I did experience two errors when running the scylladb-php-driver build.

> CMake Error at CMakeLists.txt:52 (find_package):
> By not providing "FindSanitizers.cmake" in CMAKE_MODULE_PATH this project
> has asked CMake to find a package configuration file provided by
> "Sanitizers", but CMake did not find one.

> [1/132] Building CXX object CMakeFiles/ext_scylladb.dir/php_driver.cpp.o
> FAILED: CMakeFiles/ext_scylladb.dir/php_driver.cpp.o
> /usr/bin/c++ -DCOMPILE_DL_CASSANDRA -DHAVE_CONFIG_H -DRELEASE -Dext_scylladb_EXPORTS -I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib -I/root/scylladb-php-driver/include -I/root/scylladb-php-driver/out/Release -I/root/scylladb-php-driver -O3 -DNDEBUG -std=gnu++20 -fPIC -fPIC -Wall -Wextra -Wno-long-long -Wno-deprecated-declarations -Wno-unused-parameter -Wno-unused-result -Wno-variadic-macros -Wno-format -pthread -O3 -msse2 -msse3 -msse4.1 -msse4.2 -MD -MT CMakeFiles/ext_scylladb.dir/php_driver.cpp.o -MF CMakeFiles/ext_scylladb.dir/php_driver.cpp.o.d -o CMakeFiles/ext_scylladb.dir/php_driver.cpp.o -c /root/scylladb-php-driver/php_driver.cpp
> c++: error: unrecognized command-line option ‘-msse2’
> c++: error: unrecognized command-line option ‘-msse3’
> c++: error: unrecognized command-line option ‘-msse4.1’
> c++: error: unrecognized command-line option ‘-msse4.2’ 

I worked around these errors by making two modifications to the repo.

Edit CMakeLists.txt and comment out line 52
```
#find_package(Sanitizers REQUIRED)
```

Edit cmake/TargetOptimizations.cmake and comment out line 16
```
#target_compile_options(${target} PRIVATE -msse2 -msse3 -msse4.1 -msse4.2)
```

I did also build the libscylla-cpp-driver, and libscylla-cpp-driver.so.2 does exist and is in the env PATH.

```
root@alpha-rama-ubs-vm:~# find /usr -name libscylla-cpp-driver*
/usr/local/lib/aarch64-linux-gnu/libscylla-cpp-driver_static.a
/usr/local/lib/aarch64-linux-gnu/libscylla-cpp-driver.so.2
/usr/local/lib/aarch64-linux-gnu/libscylla-cpp-driver.so.2.16.2
/usr/local/lib/aarch64-linux-gnu/libscylla-cpp-driver.so
/usr/local/lib/libscylla-cpp-driver.so.2
/usr/local/lib/libscylla-cpp-driver.so
```

```
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/usr/local/lib/aarch64-linux-gnu
```

The complete output of the build:
```
root@alpha-rama-ubs-vm:~/scylladb-php-driver# cmake --preset Release  && cd out/Release && sudo ninja install
Preset CMake variables:

  CMAKE_BUILD_TYPE="Release"
  CMAKE_INSTALL_PREFIX="/root/scylladb-php-driver/out/install/Release"
  PHP_SCYLLADB_USE_LIBCASSANDRA="OFF"
  SANITIZE_ADDRESS="OFF"
  SANITIZE_UNDEFINED="OFF"

-- Searching php-config PHP-VERSION: 8.3
-- Found php-config: /usr/bin/php-config
-- LTO is supported and enabled
-- LTO is supported and enabled
-- LTO is supported and enabled
-- LTO is supported and enabled
-- LTO is supported and enabled
-- LTO is supported and enabled
-- LTO is supported and enabled
-- LTO is supported and enabled
-- LTO is supported and enabled
-- LTO is supported and enabled
-- LTO is supported and enabled
-- LTO is supported and enabled
-- LTO is supported and enabled
-- Configuring done (2.0s)
-- Generating done (0.0s)
-- Build files have been written to: /root/scylladb-php-driver/out/Release
[131/132] Linking CXX shared library cassandra.so
/root/scylladb-php-driver/util/src/inet.cpp:33:6: warning: type ‘token_type’ violates the C++ One Definition Rule [-Wodr]
   33 | enum token_type { TOKEN_END = 0, TOKEN_COLON, TOKEN_DOT, TOKEN_HEX, TOKEN_DEC, TOKEN_ILLEGAL };
      |      ^
/root/scylladb-php-driver/util/src/types.cpp:875:6: note: an enum with different value name is defined in another translation unit
  875 | enum token_type {
      |      ^
/root/scylladb-php-driver/util/src/inet.cpp:33:19: note: name ‘TOKEN_END’ differs from name ‘TOKEN_ILLEGAL’ defined in another translation unit
   33 | enum token_type { TOKEN_END = 0, TOKEN_COLON, TOKEN_DOT, TOKEN_HEX, TOKEN_DEC, TOKEN_ILLEGAL };
      |                   ^
/root/scylladb-php-driver/util/src/types.cpp:876:3: note: mismatching definition
  876 |   TOKEN_ILLEGAL = 0,
      |   ^
/root/scylladb-php-driver/util/src/inet.cpp:35:6: warning: type ‘parser_state’ violates the C++ One Definition Rule [-Wodr]
   35 | enum parser_state {
      |      ^
/root/scylladb-php-driver/util/src/types.cpp:885:6: note: an enum with different value name is defined in another translation unit
  885 | enum parser_state { STATE_CLASS = 0, STATE_AFTER_CLASS, STATE_AFTER_PARENS, STATE_END };
      |      ^
/root/scylladb-php-driver/util/src/inet.cpp:36:3: note: name ‘STATE_START’ differs from name ‘STATE_CLASS’ defined in another translation unit
   36 |   STATE_START = 0,
      |   ^
/root/scylladb-php-driver/util/src/types.cpp:885:21: note: mismatching definition
  885 | enum parser_state { STATE_CLASS = 0, STATE_AFTER_CLASS, STATE_AFTER_PARENS, STATE_END };
      |                     ^
lto-wrapper: warning: using serial compilation of 6 LTRANS jobs
/root/scylladb-php-driver/util/src/inet.cpp: In function ‘php_driver_parse_ip_address’:
/root/scylladb-php-driver/util/src/inet.cpp:383:26: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  383 |         address[src_pos] = 0;
      |                          ^
/root/scylladb-php-driver/util/src/inet.cpp:132:16: note: at offset [-2147483648, -2] into destination object ‘address’ of size 16
  132 |   cass_uint8_t address[CASS_INET_V6_LENGTH] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
      |                ^
/root/scylladb-php-driver/util/src/inet.cpp:383:26: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  383 |         address[src_pos] = 0;
      |                          ^
/root/scylladb-php-driver/util/src/inet.cpp:132:16: note: at offset [-2147483648, -3] into destination object ‘address’ of size 16
  132 |   cass_uint8_t address[CASS_INET_V6_LENGTH] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
      |                ^
/root/scylladb-php-driver/util/src/inet.cpp:383:26: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  383 |         address[src_pos] = 0;
      |                          ^
/root/scylladb-php-driver/util/src/inet.cpp:132:16: note: at offset [-2147483648, -4] into destination object ‘address’ of size 16
  132 |   cass_uint8_t address[CASS_INET_V6_LENGTH] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
      |                ^
/root/scylladb-php-driver/util/src/inet.cpp:383:26: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  383 |         address[src_pos] = 0;
      |                          ^
/root/scylladb-php-driver/util/src/inet.cpp:132:16: note: at offset [-2147483648, -5] into destination object ‘address’ of size 16
  132 |   cass_uint8_t address[CASS_INET_V6_LENGTH] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
      |                ^
/root/scylladb-php-driver/util/src/inet.cpp:383:26: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  383 |         address[src_pos] = 0;
      |                          ^
/root/scylladb-php-driver/util/src/inet.cpp:132:16: note: at offset [-2147483648, -6] into destination object ‘address’ of size 16
  132 |   cass_uint8_t address[CASS_INET_V6_LENGTH] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
      |                ^
/root/scylladb-php-driver/util/src/inet.cpp:383:26: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  383 |         address[src_pos] = 0;
      |                          ^
/root/scylladb-php-driver/util/src/inet.cpp:132:16: note: at offset [-2147483648, -7] into destination object ‘address’ of size 16
  132 |   cass_uint8_t address[CASS_INET_V6_LENGTH] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
      |                ^
/root/scylladb-php-driver/util/src/inet.cpp:383:26: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  383 |         address[src_pos] = 0;
      |                          ^
/root/scylladb-php-driver/util/src/inet.cpp:132:16: note: at offset [-2147483648, -1] into destination object ‘address’ of size 16
  132 |   cass_uint8_t address[CASS_INET_V6_LENGTH] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
      |                ^
[131/132] Install the project...
-- Install configuration: "Release"
-- Installing: /usr/lib/php/20210902/cassandra.so
-- Set non-toolchain portion of runtime path of "/usr/lib/php/20210902/cassandra.so" to ""
-- Installing: /root/scylladb-php-driver/out/install/Release/lib/cassandra.so
-- Set non-toolchain portion of runtime path of "/root/scylladb-php-driver/out/install/Release/lib/cassandra.so" to ""
-- Installing: /usr/include/php/20210902/ext/cassandra/include/api.h
-- Installing: /usr/include/php/20210902/ext/cassandra/include/php_driver.h
-- Installing: /usr/include/php/20210902/ext/cassandra/include/php_driver_globals.h
-- Installing: /usr/include/php/20210902/ext/cassandra/include/php_driver_types.h
-- Installing: /usr/include/php/20210902/ext/cassandra/include/DateTime/Date.h
-- Installing: /usr/include/php/20210902/ext/cassandra/include/RetryPolicy/RetryPolicy.h
-- Installing: /usr/include/php/20210902/ext/cassandra/out/Release/config.h
```

# Questions

My ultimate goal is to setup a dev environment for a PHP project that needs to utilize a ScyllaDB server as a data source. My preferred work station is a Mac. 

- Is it possible to build the driver on ARM64 Architecture?
- If so, what should I try to fix the problem?
- How else should I have dealt with the build errors?
- Is there a better approach for a PHP project that needs to utilize a ScyllaDB server?
- How can I help - Testing, QA, Documentation?

# Thanks

I am grateful to everyone who has contributed to this project.

And I would greatly appreciate any advice.

**Comment by CodeLieutenant** [comment]
First of all, thank you for such a detailed set of questions and output, this means a lot 😄.

> Is it possible to build the driver on ARM64 Architecture?

Currently I don't know, I'm not running any ARM based processor and Github Actions are not giving any offerings regarding to ARM and Apple Mac M Series. Most of the deployments on the server are X86_64 Architecture and that's what we're optimizing for. 

> If so, what should I try to fix the problem?

YES, problem for me is to support something officially when I don't any hardware to test it ( Yes running AWS based Arm instances might help resolve the issue for ARM, but they are expensive and offer no developer experience to debug... could be done, but for now and me personally it's not a viable solution ).

If you are willing to support ARM64 and have some knowledge on how PHP Extensions work (even if you don't, we can make something out), SEND A PULL REQUEST PLEASE.

> How else should I have dealt with the build errors?
YOU NEED GREAT, Looks like this driver only works on X86_64 machines.

> Is there a better approach for a PHP project that needs to utilize a ScyllaDB server?

Currently NO. We are in process of rewriting the OLD Codebase to RUST on V2 Branch which will offer (we hope) same performance and better DX. There is no SCYLLADB Bindings for PHP Except this one (as far as I know -> search many times, this is why company I work for, forked it and started the maintenance)

> How can I help - Testing, QA, Documentation?

YES, Everything
Open and issue, send pull request or like you did here, open a discussions, I'm willing to make this driver better (when I have time to work on it -> I know it looks like driver is not maintained but most of the work that will be in the driver is done [here](https://github.com/dmalusev/phper) -> in order to rewrite the driver to rust, GOOD PHP BINDING ARE NEEDED)



I'll try to fix these issues, looks like a trivial fix, and issue is WITH CMakeLists

**Comment by latublu** [comment]
Thanks for replying so quickly. 

After a lot of searching, I wasn't able to find another PHP driver in active development.

I will fork the repo and do what I can.

Have you submitted the project for sponsorship?
https://github.com/sponsors

**Comment by CodeLieutenant** [comment]
> I will fork the repo and do what I can.

Nice thank you for the future contribution.

> Have you submitted the project for sponsorship?

I have not. I'm not the owner of the organization, just doing it here with @DanielHe4rt 
Really dont know if GitHub sponsors are supported in Serbia, haven't checked in a while, submitted a request for myself a long time ago and totally forgot about it, thanks for the reminder, I'll try to set it up in the near future.

If you have any questions or need any help, ping me and open Issue/Discussion
